File: Qt_turtle_service/demo/ROS_demo/tf_demo/src/dynamic_listen.cpp
#include "ros/ros.h"
#include "tf2_ros/buffer.h"
#include "tf2_ros/transform_listener.h"
#include "geometry_msgs/PointStamped.h"
#include "tf2_geometry_msgs/tf2_geometry_msgs.h"

int main(int argc, char *argv[])
{
    setlocale(LC_ALL, "");
    ros::init(argc, argv, "dynamic_tf_sub");
    ros::NodeHandle nh;
    

    tf2_ros::Buffer buffer;
    tf2_ros::TransformListener listener(buffer);

    ros::Rate r(1);

    while (ros::ok())
    {
        geometry_msgs::PointStamped absolute_points;
        absolute_points.header.frame_id = "turtle1";
        absolute_points.header.stamp = ros::Time();
        absolute_points.point.x = 1;
        absolute_points.point.y = 1;
        absolute_points.point.z = 0;

        try
        {
            geometry_msgs::PointStamped point_base;
            point_base = buffer.transform(absolute_points, "world");

            ROS_INFO("转换后的数据:(%.2f,%.2f,%.2f),参考的坐标系是:%s",
                        point_base.point.x,
                        point_base.point.y,
                        point_base.point.z,
                        point_base.header.frame_id.c_str());
        }
        catch(const std::exception& e)
        {
            // std::cerr << e.what() << '\n';
            ROS_INFO("程序异常.....");
        }
        r.sleep();
        ros::spinOnce();
        
    }
    


    return 0;
}

File: ROS_demo/tf_demo/scripts/dynamic_listen.py
#!/usr/bin/env python
# coding:utf-8

from logging import exception
import tf.transformations
import rospy
from turtlesim.msg import Pose
import tf2_ros
from tf2_geometry_msgs import PointStamped


if __name__ == "__main__":
    node_name = "dynamic_listen"
    rospy.init_node(node_name)
    
    buffer = tf2_ros.Buffer()
    listener = tf2_ros.TransformListener(buffer)

    rate = rospy.Rate(1)

    while not rospy.is_shutdown():
        point_source = PointStamped()
        point_source.header.frame_id = "turtle1"
        point_source.header.stamp = rospy.Time()
        point_source.point.x = 10
        point_source.point.y = 2
        point_source.point.z = 3
        try:
            point_target = buffer.transform(point_source,"world",rospy.Duration(1))
            
            rospy.loginfo("转换结果:x = %.2f, y = %.2f, z = %.2f",
                            point_target.point.x,
                            point_target.point.y,
                            point_target.point.z)

        except exception as e:
            rospy.logerr("异常: %s", e)

        rate.sleep()

    